Goobi/plugins/opac/ModsPlugin/org/kitodo/production/plugin/CataloguePlugin/ModsPlugin/ModsPlugin.java

```diff
@@ -735,32 +783,61 @@ private String retrieveParentRecord(Document doc, long timeout) throws JDOMExcep
-     * sections and a METS structural map constructed from the given List
-     * <Element> 'dmdElements' and List<String> 'documentTypes'.
+     * Retrieves and returns the child elements of the document with the given ID 'documentId'
+     * from the Kalliope SRU interface.
      *
-     * @param dmdElements
-     * @param documentTypes
-     * @return
+     * @param documentId
+     * @param timeout
+     * @return the list of child elements
+     * @throws JDOMException
      */
-    private static Document createMetsContainer(List<Element> dmdElements, List<String> documentTypes) {
-        assert (dmdElements.size() == documentTypes.size());
-
-        Document metsDocument = createMetsDocument();
+    private List<Element> retrieveChildDocuments(String documentId, long timeout) throws JDOMException {
+        List<Element> childDocuments = new LinkedList<Element>();
+        Query childrenQuery = new Query("context.ead.id:" + documentId);
+        String allChildren = client.retrieveModsRecord(childrenQuery.getQueryUrl(), timeout);
 
-        for (Element dmdSection : dmdElements) {
-            metsDocument = addSectionToMETSDocument(metsDocument, dmdSection);
+        SAXBuilder sb = new SAXBuilder();
+        try {
+            Document childrenDoc = sb.build(new StringReader(allChildren));
+            for (Object child : modsXPath.selectNodes(childrenDoc)) {
+                childDocuments.add((Element) child);
+            }
+        } catch (IOException e) {
+            logger.error("Unable to import child elements of document with given ID.");
         }
+        return childDocuments;
+    }
 
-        Element structMap = createMETSStructureMap(METS_LOGICAL);
-
-        Element currentParent = structMap;
-        for (int i = dmdElements.size() - 1; i >= 0; i--) {
-            currentParent = addDivToMETSStructureMap(currentParent, dmdElements.get(i), documentTypes.get(i));
+    /**
+     * Clones the given document 'doc' and removes all children from the cloned documen
+     * except the child at position 'index'. Returns the cloned document with one remaining
+     * child.
+     *
+     * @param doc
+     * @param index
+     * @return document with one remaining child
+     */
+    @SuppressWarnings("unchecked")
+    private Document removeAllChildrenButOne(Document doc, int index) {
+        try {
+            Document clone = (Document) doc.clone();
+            Element remainingChild = null;
+            ArrayList<Element> recordNodes = (ArrayList<Element>) srwRecordXPath.selectNodes(clone);
+
+            for (int i = 0; i < recordNodes.size(); i++) {
+                Element currentRecord = recordNodes.get(i);
+                if (i != index) {
+                    currentRecord.detach();
+                } else {
+                    remainingChild = currentRecord;
+                }
+            }
+            return remainingChild.getDocument();
```

Comment:
`remainingChild` could be `null` and a NPE could be thrown. This should be fixed / avoided.